File: src/Core.Go/yaml/yaml.go
package yaml

import (
	"errors"
	"regexp"
	"strings"
)

type ScalarKind string

const (
	Null  ScalarKind = "Null"
	Bool  ScalarKind = "Bool"
	Int   ScalarKind = "Int"
	Float ScalarKind = "Float"
	Str   ScalarKind = "Str"
)

type ScalarStyle string

const (
	Plain        ScalarStyle = "Plain"
	SingleQuoted ScalarStyle = "SingleQuoted"
	DoubleQuoted ScalarStyle = "DoubleQuoted"
)

type YamlEvent struct {
	E     string       `json:"e"`
	Raw   *string      `json:"raw,omitempty"`
	Kind  *ScalarKind  `json:"kind,omitempty"`
	Style *ScalarStyle `json:"style,omitempty"`
}

var (
	nullLiterals = map[string]bool{"": true, "~": true, "null": true, "Null": true, "NULL": true}
	boolLiterals = map[string]bool{"true": true, "True": true, "TRUE": true, "false": true, "False": true, "FALSE": true}
	intRe        = regexp.MustCompile(`^-?[0-9]+$`)
	floatRe      = regexp.MustCompile(`^-?[0-9]+\.[0-9]+([eE][-+]?[0-9]+)?$`)
)

func resolvePlainKind(raw string) ScalarKind {
	if nullLiterals[raw] {
		return Null
	}
	if boolLiterals[raw] {
		return Bool
	}
	if intRe.MatchString(raw) {
		return Int
	}
	if floatRe.MatchString(raw) {
		return Float
	}
	return Str
}

var unsupportedValueStart = map[rune]bool{
	'&': true, '*': true, '!': true, '{': true, '[': true, '|': true, '>': true,
}

func isSpace(r rune) bool {
	return r == ' ' || r == '\t'
}

func leftTrim(s string) string {
	runes := []rune(s)
	i := 0
	for i < len(runes) && isSpace(runes[i]) {
		i++
	}
	return string(runes[i:])
}

func stripTrailingComment(token string) string {
	runes := []rune(token)
	for i := 0; i < len(runes); i++ {
		if runes[i] == '#' && i > 0 && isSpace(runes[i-1]) {
			return string(runes[:i])
		}
	}
	return token
}

func decodeSingleQuoted(token string) (string, error) {
	var sb strings.Builder
	runes := []rune(token)
	for i := 1; i < len(runes); i++ {
		r := runes[i]
		if r == '\'' {
			if i+1 < len(runes) && runes[i+1] == '\'' {
				sb.WriteRune('\'')
				i++
				continue
			}
			return sb.String(), nil
		}
		sb.WriteRune(r)
	}
	return "", errors.New("UnterminatedQuote")
}

func decodeDoubleQuoted(token string) (string, error) {
	var sb strings.Builder
	runes := []rune(token)
	for i := 1; i < len(runes); i++ {
		r := runes[i]
		if r == '"' {
			return sb.String(), nil
		}
		if r == '\\' {
			if i+1 >= len(runes) {
				return "", errors.New("UnterminatedQuote")
			}
			next := runes[i+1]
			switch next {
			case '\\':
				sb.WriteRune('\\')
			case '"':
				sb.WriteRune('"')
			case 'n':
				sb.WriteRune('\n')
			case 't':
				sb.WriteRune('\t')
			case 'r':
				sb.WriteRune('\r')
			case '0':
				sb.WriteRune('\x00')
			case '/':
				sb.WriteRune('/')
			default:
				return "", errors.New("UnexpectedCharacter")
			}
			i++
			continue
		}
		sb.WriteRune(r)
	}
	return "", errors.New("UnterminatedQuote")
}

type parsedScalar struct {
	raw   string
	kind  ScalarKind
	style ScalarStyle
}

func parseValue(token string) (parsedScalar, error) {
	if len(token) == 0 {
		return parsedScalar{raw: "", kind: Null, style: Plain}, nil
	}
	runes := []rune(token)
	first := runes[0]
	if first == '"' {
		dec, err := decodeDoubleQuoted(token)
		if err != nil {
			return parsedScalar{}, err
		}
		return parsedScalar{raw: dec, kind: Str, style: DoubleQuoted}, nil
	}
	if first == '\'' {
		dec, err := decodeSingleQuoted(token)
		if err != nil {
			return parsedScalar{}, err
		}
		return parsedScalar{raw: dec, kind: Str, style: SingleQuoted}, nil
	}
	if unsupportedValueStart[first] {
		return parsedScalar{}, errors.New("UnsupportedConstruct")
	}
	raw := strings.TrimRight(stripTrailingComment(token), " \t\r\n")
	return parsedScalar{raw: raw, kind: resolvePlainKind(raw), style: Plain}, nil
}

type contentLine struct {
	indent int
	text   string
}

func toContentLines(text string) ([]contentLine, error) {
	var out []contentLine
	lines := strings.Split(text, "\n")
	for _, raw := range lines {
		raw = strings.TrimSuffix(raw, "\r")

		indent := 0
		sawTab := false
		runes := []rune(raw)
	loop:
		for indent < len(runes) {
			ch := runes[indent]
			switch ch {
			case ' ':
				indent++
			case '\t':
				sawTab = true
				indent++
			default:
				break loop
			}
		}

		body := string(runes[indent:])
		if len(body) == 0 {
			continue
		}
		if body[0] == '#' {
			continue
		}
		if sawTab {
			return nil, errors.New("TabIndentation")
		}

		out = append(out, contentLine{indent: indent, text: body})
	}
	return out, nil
}

func isDocumentMarker(text string) bool {
	t := strings.TrimRight(text, " \t\r\n")
	return t == "---" || t == "..." || strings.HasPrefix(t, "--- ") || strings.HasPrefix(t, "... ")
}

type mappingEntry struct {
	key   string
	value *string
}

func splitMappingEntry(text string) *mappingEntry {
	inSingle := false
	inDouble := false
	runes := []rune(text)
	for i := 0; i < len(runes); i++ {
		ch := runes[i]
		if inSingle {
			if ch == '\'' {
				inSingle = false
			}
			continue
		}
		if inDouble {
			if ch == '\\' {
				i++
				continue
			}
			if ch == '"' {
				inDouble = false
			}
			continue
		}
		if ch == '\'' {
			inSingle = true
			continue
		}
		if ch == '"' {
			inDouble = true
			continue
		}
		if ch == '#' && i > 0 && isSpace(runes[i-1]) {
			return nil
		}
		if ch == ':' && (i+1 >= len(runes) || runes[i+1] == ' ' || runes[i+1] == '\t') {
			key := strings.TrimRight(string(runes[:i]), " \t")
			var val *string
			if i+1 < len(runes) {
				trimmed := leftTrim(string(runes[i+1:]))
				if len(trimmed) > 0 && !strings.HasPrefix(trimmed, "#") {
					val = &trimmed
				}
			}
			return &mappingEntry{key: key, value: val}
		}
	}
	return nil
}

type frame struct {
	indent int
	kind   string // "Mapping" or "Sequence"
}

func ReadEvents(text string) ([]YamlEvent, error) {
	lines, err := toContentLines(text)
	if err != nil {
		return nil, err
	}

	events := []YamlEvent{{E: "StreamStart"}}
	var stack []frame

	pushContainer := func(indent int, kind string) {
		stack = append(stack, frame{indent: indent, kind: kind})
		if kind == "Mapping" {
			events = append(events, YamlEvent{E: "MappingStart"})
		} else {
			events = append(events, YamlEvent{E: "SequenceStart"})
		}
	}

	popGreaterThan := func(indent int) {
		for len(stack) > 0 && stack[len(stack)-1].indent > indent {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if top.kind == "Mapping" {
				events = append(events, YamlEvent{E: "MappingEnd"})
			} else {
				events = append(events, YamlEvent{E: "SequenceEnd"})
			}
		}
	}

	childIndentAt := func(idx int) int {
		if idx+1 < len(lines) {
			return lines[idx+1].indent
		}
		return -1
	}

	peekChildKind := func(idx int, childIndent int) string {
		if idx+1 >= len(lines) {
			return "Mapping"
		}
		next := lines[idx+1]
		if next.indent != childIndent {
			return "Mapping"
		}
		if next.text == "-" || strings.HasPrefix(next.text, "- ") {
			return "Sequence"
		}
		return "Mapping"
	}

	emitKey := func(key string) error {
		parsed, err := parseValue(key)
		if err != nil {
			return err
		}
		kindStr := Str
		events = append(events, YamlEvent{E: "Scalar", Raw: &parsed.raw, Kind: &kindStr, Style: &parsed.style})
		return nil
	}

	emitNull := func() {
		rawVal := ""
		kindVal := Null
		styleVal := Plain
		events = append(events, YamlEvent{E: "Scalar", Raw: &rawVal, Kind: &kindVal, Style: &styleVal})
	}

	emitValue := func(value string) error {
		parsed, err := parseValue(value)
		if err != nil {
			return err
		}
		events = append(events, YamlEvent{E: "Scalar", Raw: &parsed.raw, Kind: &parsed.kind, Style: &parsed.style})
		return nil
	}

	emitValueOrEmptyFlow := func(value string) error {
		token := strings.TrimRight(stripTrailingComment(value), " \t\r\n")
		switch token {
		case "{}":
			events = append(events, YamlEvent{E: "MappingStart"})
			events = append(events, YamlEvent{E: "MappingEnd"})
			return nil
		case "[]":
			events = append(events, YamlEvent{E: "SequenceStart"})
			events = append(events, YamlEvent{E: "SequenceEnd"})
			return nil
		default:
			return emitValue(value)
		}
	}

	for li := 0; li < len(lines); li++ {
		line := lines[li]
		body := line.text
		indent := line.indent

		if isDocumentMarker(body) {
			return nil, errors.New("UnsupportedConstruct")
		}

		isSeqItem := body == "-" || strings.HasPrefix(body, "- ")

		if len(stack) == 0 {
			kind := "Mapping"
			if isSeqItem {
				kind = "Sequence"
			}
			pushContainer(indent, kind)
		} else if indent <= stack[len(stack)-1].indent {
			popGreaterThan(indent)
			if len(stack) == 0 {
				return nil, errors.New("UnexpectedIndent")
			}
			top := stack[len(stack)-1]
			if top.indent != indent {
				return nil, errors.New("UnexpectedIndent")
			}
			if isSeqItem != (top.kind == "Sequence") {
				return nil, errors.New("UnexpectedIndent")
			}
		}

		if isSeqItem {
			var afterDash string
			if body == "-" {
				afterDash = ""
			} else {
				afterDash = body[2:]
			}
			itemContent := leftTrim(afterDash)

			if len(itemContent) == 0 || strings.HasPrefix(itemContent, "#") {
				childIndent := childIndentAt(li)
				if childIndent > indent {
					pushContainer(childIndent, peekChildKind(li, childIndent))
				} else {
					emitNull()
				}
				continue
			}

			innerEntry := splitMappingEntry(itemContent)
			if innerEntry != nil {
				mapIndent := indent + (len(body) - len(itemContent))
				pushContainer(mapIndent, "Mapping")
				err := emitKey(innerEntry.key)
				if err != nil {
					return nil, err
				}
				if innerEntry.value != nil {
					err = emitValueOrEmptyFlow(*innerEntry.value)
					if err != nil {
						return nil, err
					}
				} else {
					childIndent := childIndentAt(li)
					if childIndent > mapIndent {
						pushContainer(childIndent, peekChildKind(li, childIndent))
					} else {
						emitNull()
					}
				}
				continue
			}

			err := emitValueOrEmptyFlow(itemContent)
			if err != nil {
				return nil, err
			}
			continue
		}

		entry := splitMappingEntry(body)
		if entry == nil {
			return nil, errors.New("UnsupportedConstruct")
		}
		err := emitKey(entry.key)
		if err != nil {
			return nil, err
		}
		if entry.value != nil {
			err = emitValueOrEmptyFlow(*entry.value)
			if err != nil {
				return nil, err
			}
		} else {
			childIndent := childIndentAt(li)
			if childIndent > indent {
				pushContainer(childIndent, peekChildKind(li, childIndent))
			} else {
				emitNull()
			}
		}
	}

	for len(stack) > 0 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if top.kind == "Mapping" {
			events = append(events, YamlEvent{E: "MappingEnd"})
		} else {
			events = append(events, YamlEvent{E: "SequenceEnd"})
		}
	}
	events = append(events, YamlEvent{E: "StreamEnd"})
	return events, nil
}

File: src/Core.Python/src/zeta/yaml.py
# yaml.py — Python port of custom config-friendly YAML events scanner.
import re
from typing import Any

INT_RE = re.compile(r"^-?[0-9]+$")
FLOAT_RE = re.compile(r"^-?[0-9]+\.[0-9]+([eE][-+]?[0-9]+)?$")

NULL_LITERALS = {"", "~", "null", "Null", "NULL"}
BOOL_LITERALS = {"true", "True", "TRUE", "false", "False", "FALSE"}
UNSUPPORTED_VALUE_START = {"&", "*", "!", "{", "[", "|", ">"}


class DeclineError(Exception):
    def __init__(self, feedback: str):
        super().__init__(feedback)
        self.feedback = feedback


def resolve_plain_kind(raw: str) -> str:
    if raw in NULL_LITERALS:
        return "Null"
    if raw in BOOL_LITERALS:
        return "Bool"
    if INT_RE.match(raw):
        return "Int"
    if FLOAT_RE.match(raw):
        return "Float"
    return "Str"


def is_space(ch: str) -> bool:
    return ch == " " or ch == "\t"


def left_trim(s: str) -> str:
    i = 0
    while i < len(s) and is_space(s[i]):
        i += 1
    return s[i:]


def strip_trailing_comment(token: str) -> str:
    for i in range(len(token)):
        if token[i] == "#" and i > 0 and is_space(token[i - 1]):
            return token[:i]
    return token


def decode_single_quoted(token: str) -> str:
    out: list[str] = []
    i = 1
    while i < len(token):
        ch = token[i]
        if ch == "'":
            if i + 1 < len(token) and token[i + 1] == "'":
                out.append("'")
                i += 2
                continue
            return "".join(out)
        out.append(ch)
        i += 1
    raise DeclineError("UnterminatedQuote")


def decode_double_quoted(token: str) -> str:
    out: list[str] = []
    i = 1
    while i < len(token):
        ch = token[i]
        if ch == '"':
            return "".join(out)
        if ch == "\\":
            if i + 1 >= len(token):
                raise DeclineError("UnterminatedQuote")
            next_ch = token[i + 1]
            if next_ch == "\\":
                out.append("\\")
            elif next_ch == '"':
                out.append('"')
            elif next_ch == "n":
                out.append("\n")
            elif next_ch == "t":
                out.append("\t")
            elif next_ch == "r":
                out.append("\r")
            elif next_ch == "0":
                out.append("\x00")
            elif next_ch == "/":
                out.append("/")
            else:
                raise DeclineError("UnexpectedCharacter")
            i += 2
            continue
        out.append(ch)
        i += 1
    raise DeclineError("UnterminatedQuote")


def parse_value(token: str):
    if not token:
        return {"raw": "", "kind": "Null", "style": "Plain"}
    first = token[0]
    if first == '"':
        return {
            "raw": decode_double_quoted(token),
            "kind": "Str",
            "style": "DoubleQuoted",
        }
    if first == "'":
        return {
            "raw": decode_single_quoted(token),
            "kind": "Str",
            "style": "SingleQuoted",
        }
    if first in UNSUPPORTED_VALUE_START:
        raise DeclineError("UnsupportedConstruct")
    raw = strip_trailing_comment(token).rstrip(" \t\r\n")
    return {"raw": raw, "kind": resolve_plain_kind(raw), "style": "Plain"}


def to_content_lines(text: str):
    out = []
    for raw in text.split("\n"):
        raw = raw.removesuffix("\r")

        indent = 0
        saw_tab = False
        while indent < len(raw):
            ch = raw[indent]
            if ch == " ":
                indent += 1
            elif ch == "\t":
                saw_tab = True
                indent += 1
            else:
                break

        body = raw[indent:]
        if not body:
            continue
        if body[0] == "#":
            continue
        if saw_tab:
            raise DeclineError("TabIndentation")

        out.append({"indent": indent, "text": body})
    return out


def is_document_marker(text: str) -> bool:
    t = text.rstrip(" \t\r\n")
    return t in ("---", "...") or t.startswith(("--- ", "... "))


def split_mapping_entry(text: str):
    in_single = False
    in_double = False
    i = 0
    while i < len(text):
        ch = text[i]
        if in_single:
            if ch == "'":
                in_single = False
            i += 1
            continue
        if in_double:
            if ch == "\\":
                i += 2
                continue
            if ch == '"':
                in_double = False
            i += 1
            continue
        if ch == "'":
            in_single = True
            i += 1
            continue
        if ch == '"':
            in_double = True
            i += 1
            continue
        if ch == "#" and i > 0 and is_space(text[i - 1]):
            return None
        if ch == ":" and (
            i + 1 >= len(text) or text[i + 1] == " " or text[i + 1] == "\t"
        ):
            key = text[:i].rstrip(" \t")
            val = None
            if i + 1 < len(text):
                trimmed = left_trim(text[i + 1 :])
                if trimmed and not trimmed.startswith("#"):
                    val = trimmed
            return {"key": key, "value": val}
        i += 1
    return None


def read_events(text: str):
    lines = to_content_lines(text)
    events = [{"e": "StreamStart"}]
    stack: list[dict[str, Any]] = []

    def push_container(indent: int, kind: str):
        stack.append({"indent": indent, "kind": kind})
        events.append({"e": "MappingStart" if kind == "Mapping" else "SequenceStart"})

    def pop_greater_than(indent: int):
        while stack and stack[-1]["indent"] > indent:
            top = stack.pop()
            events.append(
                {"e": "MappingEnd" if top["kind"] == "Mapping" else "SequenceEnd"}
            )

    def child_indent_at(idx: int) -> int:
        if idx + 1 < len(lines):
            return lines[idx + 1]["indent"]
        return -1

    def peek_child_kind(idx: int, child_indent: int) -> str:
        if idx + 1 >= len(lines):
            return "Mapping"
        next_line = lines[idx + 1]
        if next_line["indent"] != child_indent:
            return "Mapping"
        if next_line["text"] == "-" or next_line["text"].startswith("- "):
            return "Sequence"
        return "Mapping"

    def emit_key(key: str):
        parsed = parse_value(key)
        events.append(
            {
                "e": "Scalar",
                "raw": parsed["raw"],
                "kind": "Str",
                "style": parsed["style"],
            }
        )

    def emit_null():
        events.append({"e": "Scalar", "raw": "", "kind": "Null", "style": "Plain"})

    def emit_value(value: str):
        parsed = parse_value(value)
        events.append(
            {
                "e": "Scalar",
                "raw": parsed["raw"],
                "kind": parsed["kind"],
                "style": parsed["style"],
            }
        )

    def emit_value_or_empty_flow(value: str):
        token = strip_trailing_comment(value).rstrip(" \t\r\n")
        if token == "{}":
            events.append({"e": "MappingStart"})
            events.append({"e": "MappingEnd"})
        elif token == "[]":
            events.append({"e": "SequenceStart"})
            events.append({"e": "SequenceEnd"})
        else:
            emit_value(value)

    for li in range(len(lines)):
        line = lines[li]
        body = line["text"]
        indent = line["indent"]

        if is_document_marker(body):
            raise DeclineError("UnsupportedConstruct")

        is_seq_item = body == "-" or body.startswith("- ")

        if not stack:
            push_container(indent, "Sequence" if is_seq_item else "Mapping")
        elif indent <= stack[-1]["indent"]:
            pop_greater_than(indent)
            if not stack:
                raise DeclineError("UnexpectedIndent")
            top = stack[-1]
            if top["indent"] != indent:
                raise DeclineError("UnexpectedIndent")
            if is_seq_item != (top["kind"] == "Sequence"):
                raise DeclineError("UnexpectedIndent")

        if is_seq_item:
            after_dash = "" if body == "-" else body[2:]
            item_content = left_trim(after_dash)

            if not item_content or item_content.startswith("#"):
                child_indent = child_indent_at(li)
                if child_indent > indent:
                    push_container(child_indent, peek_child_kind(li, child_indent))
                else:
                    emit_null()
                continue

            inner_entry = split_mapping_entry(item_content)
            if inner_entry is not None:
                map_indent = indent + (len(body) - len(item_content))
                push_container(map_indent, "Mapping")
                emit_key(inner_entry["key"])
                if inner_entry["value"] is not None:
                    emit_value_or_empty_flow(inner_entry["value"])
                else:
                    child_indent = child_indent_at(li)
                    if child_indent > map_indent:
                        push_container(child_indent, peek_child_kind(li, child_indent))
                    else:
                        emit_null()
                continue

            emit_value_or_empty_flow(item_content)
            continue

        entry = split_mapping_entry(body)
        if entry is None:
            raise DeclineError("UnsupportedConstruct")
        emit_key(entry["key"])
        if entry["value"] is not None:
            emit_value_or_empty_flow(entry["value"])
        else:
            child_indent = child_indent_at(li)
            if child_indent > indent:
                push_container(child_indent, peek_child_kind(li, child_indent))
            else:
                emit_null()

    while stack:
        top = stack.pop()
        events.append(
            {"e": "MappingEnd" if top["kind"] == "Mapping" else "SequenceEnd"}
        )
    events.append({"e": "StreamEnd"})
    return events
